ColorPicker/utils.js

```diff
@@ -1,35 +1,143 @@
+	let r = 0, g = 0, b = 0;
+
+	if (0 <= h && h < 60) {
+		r = c; g = x; b = 0;
+	} else if (60 <= h && h < 120) {
+		r = x; g = c; b = 0;
+	} else if (120 <= h && h < 180) {
+		r = 0; g = c; b = x;
+	} else if (180 <= h && h < 240) {
+		r = 0; g = x; b = c;
+	} else if (240 <= h && h < 300) {
+		r = x; g = 0; b = c;
+	} else if (300 <= h && h <= 360) {
+		r = c; g = 0; b = x;
+	}
+
+	return {r, g, b, m};
+};
+
+/**
+ * Converts red, green, and blue string to HEX color.
+ *
+ * @function
+ * @param {string} rgbString
+ * @returns {string|boolean} 6-digit HEX color if the input is valid, or `false` if the input is invalid
+ * @private
+ */
+const rgbStringToHex = (rgbString) => {
+	// Step 1: Check if the input starts with "rgb("
+	if (!rgbString.startsWith("rgb(")) return false;
+
+	// Step 2: Extract the numbers between "rgb(" and ")"
+	let a = rgbString.slice(4, -1).split(",").map(s => s.trim());
+
+	// Step 3: Validate that the array has exactly 3 valid RGB values (0–255)
+	if (
+		a.length !== 3 || // Ensure there are 3 values
+		!a.every(v => /^\d+$/.test(v) && +v >= 0 && +v <= 255) // Check for valid digits and range
+	) return false;
+
+	// Step 4: Convert each RGB value to a 2-digit hex string
+	return "#" + a.map(v => (+v).toString(16).padStart(2, "0")).join("").toUpperCase();
+};
+
+/**
+ * Converts red, green, and blue values to HEX color.
+ *
+ * @function
+ * @param r {number}
+ * @param g {number}
+ * @param b {number}
+ * @returns {String} 6-digit HEX color
+ * @private
+ */
+const rgbToHex = (r, g, b) => {
+	return '#' + [r, g, b].map(x => x.toString(16).padStart(2, '0')).join('');
+};
+
+/**
+ * Extracts color from gradient and converts it to HEX color.
+ *
+ * @function
+ * @param {Object} canvasRef
+ * @param {number} x
+ * @param {number} y
+ * @returns {String} 6-digit HEX color
+ * @private
+ */
+const getHexColorFromGradient = (canvasRef, x, y) => {
+	const ctx = canvasRef.current.getContext('2d');
+	const imageData = ctx.getImageData(x, y, 1, 1);
+	return rgbToHex(imageData.data[0], imageData.data[1], imageData.data[2]);
+};
+
+/**
+ * Generates complementary HEX color.
+ *
+ * @function
  * @param {String} hexColor
- * @returns {{s: number, h: number, l: number}}	HSL values
+ * @returns {String} 6-digit HEX color
+ * @private
+ */
+const generateOppositeColor = (hexColor) => {
+	hexColor = hexColor.replace('#', '');
+
+	const bigint = parseInt(hexColor, 16);
+	const r = (bigint >> 16) & 255;
+	const g = (bigint >> 8) & 255;
+	const b = bigint & 255;
+
+	const luminance = 0.2126 * r + 0.7152 * g + 0.0722 * b;
+
+	return luminance > 128 ? '#101720CC' : '#FFFFFFCC';
+};
+
+/**
+ * Converts HEX color to hue, saturation, and lightness values.
+ *
+ * @function
+ * @param {String} hexColor
+ * @returns {{h: number, s: number, l: number}}	HSL values
  * @private
  */
 const hexToHSL = (hexColor) => {
-	const hex = checkHex(hexColor);
 	// Convert hex to RGB first
 	let r = 0, g = 0, b = 0;
```

Comment:
variable initiliazer is redundant. let's define r,g,b directly in the lines below